modules/core/arrow-annotations-processor/src/main/java/arrow/generic/CoproductFileGenerator.kt

```diff
@@ -0,0 +1,194 @@
+import com.squareup.kotlinpoet.PropertySpec
+import com.squareup.kotlinpoet.TypeSpec
+import com.squareup.kotlinpoet.TypeVariableName
+import java.io.File
+import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
+
+private val genericsToClassNames = mapOf(
+        "A" to "First",
+        "B" to "Second",
+        "C" to "Third",
+        "D" to "Fourth",
+        "E" to "Fifth",
+        "F" to "Sixth",
+        "G" to "Seventh",
+        "H" to "Eighth",
+        "I" to "Ninth",
+        "J" to "Tenth",
+        "K" to "Eleventh",
+        "L" to "Twelfth",
+        "M" to "Thirteenth",
+        "N" to "Fourteenth",
+        "O" to "Fifteenth",
+        "P" to "Sixteenth",
+        "Q" to "Seventeenth",
+        "R" to "Eighteenth",
+        "S" to "Nineteenth",
+        "T" to "Twentieth",
+        "U" to "TwentyFirst",
+        "V" to "TwentySecond"
+)
+
+fun generateCoproducts(destination: File) {
+    for (size in 2 until genericsToClassNames.size + 1) {
+        val generics = genericsToClassNames.keys.toList().take(size)
+
+        FileSpec.builder("arrow.generic.coproduct$size", "Coproduct$size")
+                .apply {
+                    addCoproductClassDeclaration(generics)
+                    addCoproductOfConstructors(generics)
+                    addCopExtensionConstructors(generics)
+                    addSelectFunctions(generics)
+                    addFoldFunction(generics)
+                }
+                .build()
+                .writeTo(destination)
+    }
+}
+
+private fun FileSpec.Builder.addCoproductClassDeclaration(generics: List<String>) {
+    addType(
+            TypeSpec.classBuilder("Coproduct${generics.size}")
+                    .addModifiers(KModifier.SEALED)
+                    .addTypeVariables(generics.map { TypeVariableName(it) })
+                    .build()
+    )
+
+    for (generic in generics) {
+        addType(
+                TypeSpec.classBuilder(genericsToClassNames[generic]!!)
+                        .addModifiers(KModifier.INTERNAL, KModifier.DATA)
+                        .addTypeVariables(generics.toTypeParameters())
+                        .superclass(parameterizedCoproductNClassName(generics))
+                        .addProperty(
+                                PropertySpec.builder(generic.toLowerCase(), TypeVariableName(generic))
+                                        .initializer(generic.toLowerCase())
+                                        .build()
+                        )
+                        .primaryConstructor(
+                                FunSpec.constructorBuilder()
+                                        .addParameter(generic.toLowerCase(), TypeVariableName(generic))
+                                        .build()
+                        )
+                        .build()
+        )
+    }
+}
+
+private fun FileSpec.Builder.addCoproductOfConstructors(generics: List<String>) {
```

Comment:
please type all functions regardless of them being private or overrides. We prefer not to rely on type inference in Arrow since typed functions serve as documentation and are explicit as to what they do.

Reply:
Oh gotcha, yeah I can add the `Unit` returns on these. I suppose I could return `this` and axe the `apply` too